---
title: "Review and testing of gwasurvivr: an R package for genome-wide
survival analysis"

author:
- Nadia
- Ivy
- Gabriel
- Piotr

output:
  pdf_document: default
  html_document:
    df_print: paged
date: "2023-10-05"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

#for installing the packages

#install.packages("tinytex")
#tinytex::install_tinytex()
#if (!requireNamespace("BiocManager", quietly=TRUE))
#    install.packages("BiocManager")
#BiocManager::install("gwasurvivr", version = "devel")
#install.packages("devtools")
#devtools::install_github("kupietz/kableExtra")

library(gwasurvivr)
options("gwasurvivr.cores"=4) 
library(kableExtra)
```

## Introduction

Survival analysis has an important place in biomedical research, facilitating the exploration of time-to-event outcomes such as mortality or relapse. 

An essential component in exploring the genetic basis of diseases involves investigating Single Nucleotide Polymorphisms (SNPs). These occasional variations in a single letter of DNA can have a significant impact on susceptibility to certain diseases or on response to medical treatments. 

Integrating SNPs into survival analysis enables the discovery of genetic factors linked to time-to-event outcomes, shedding light on the genetic factors that influence disease progression and other critical events.

However, the major challenge is that our genome is made up of millions of SNPs, making large-scale survival analysis (GWAS) extremely complex. The existing software options for conducting such analyses are limited in several aspects (the need to interact with raw data, software not suited for survival analysis, and long execution times that hinder scalability). Consequently, researchers often face practical difficulties when conducting large-scale survival analyses.

gwasurvivr, an R/Bioconductor package was designed to surmount these challenges. This library offers a significant advancement in allowing researchers to perform survival analysis on large SNP datasets with remarkable efficiency and accuracy and with multiple file input formats such as VCF, IMPUTE2 or PLINK. 

In this paper, we thoroughly examine the functionalities of gwasurvivr and offer an extensive evaluation of its operational mechanisms and effectiveness in unraveling the genetic factors that influence disease survival.

## Methods

## Datasets

`gwasurvivr` provides multiple functions to run a Survival analyses. The functions
accept a file containing SNPs in different formats which is indicated by the 
respective name, e.g. `.gds` format is run by the method `gdsCoxSurv`. Each function
also needs a file which contains the meta data of patients, like sex, survival time,
vital status or age, to fit the survival model.

An example dataset is provided in the extdata folder which is automatically
installed with `gwasurvivr` containing all supported dataformats, including
`.gds`, `IMPUTE2`, `.vcf` or related formats from the Michigan or Sanger imputation
Server or the, `.bed` files or related formats generated by the programme 
`PLINK` or its successor `PLINK2`

## Results

Michigan Imputation Server pre-phases typed genotypes using HAPI-UR, SHAPEIT, or EAGLE (default is EAGLE2), imputes using Minimac3 imputation engine and outputs Blocked GNU Zip Format VCF files (`.vcf.gz`). These `vcf.gz` files are used as input for `gwasurvivr`. Process of importing `vcf.gz` and phenotype files from `gwasurvivr` is shown below.

```{r data loading, echo=T}
vcf.file <- system.file(package="gwasurvivr",
                        "extdata", 
                        "michigan.chr14.dose.vcf.gz")
pheno.fl <- system.file(package="gwasurvivr",
                        "extdata", 
                        "simulated_pheno.txt")
```

The simulated phenotype file can be represented in the table.

```{r data showcase, echo=F}
pheno.file <- read.table(pheno.fl,
                         sep=" ", 
                         header=TRUE,
                         stringsAsFactors = FALSE)
head(pheno.file) %>% kable(booktabs=F, linesep = "") %>% kable_styling(latex_options = c("HOLD_position"))
```

Now using phenotype file as covariate file and given VCF file we can run `michiganCoxSurv` wrapper for Cox regression model.

```{r cox regression, echo=T, eval=F, cache=T}
#decoding sex into binary format
pheno.file$SexFemale <- ifelse(pheno.file$sex=="female", 1L, 0L)

#running Cox regression
michiganCoxSurv(vcf.file=vcf.file,
                covariate.file=pheno.file,
                id.column="ID_2",
                time.to.event="time",
                event="event",
                covariates=c("age", "SexFemale", "DrugTxYes"),
                inter.term=NULL, #interaction term inclusion
                print.covs="only", #defines printing of covariates' statistics
                out.file="michigan_only",
                r2.filter=0.3, #imputation quality score filter
                maf.filter=0.005, #filter for minor allele frequency
                chunk.size=100, #number of variants to proceed per thread
                verbose=F,
                clusterObj=NULL) #for setting up cluster for computations
```

Functions saves the outputed model with the .coxph extension as a seperate file as well as SNPs removed (pvalues below 0.05) in the .snps_removed file.
Accessed results of the performed regression are showcased below (`print covs = "only"` was chosen as a printing option).

```{r coxph results, echo=F}
michigan_only <- read.table("michigan_only.coxph", sep="\t", header=TRUE, stringsAsFactors = FALSE) %>% t()
kable(michigan_only, linesep = "") %>% kable_styling(latex_options = c("HOLD_position"))
```

To decipher most column names and extract knowledge from the output table in the appendix section explains the meaning of certain variables.

## Conclusions

## Appendix

|   |   |
|---|---|
| RSID  | SNP ID | 	
| TYPED | Imputation status: TRUE (SNP IS TYPED)/FALSE (SNP IS IMPUTED)|
|CHR |	Chromosome number |
|POS |	Genomic Position (BP)|
|REF |	Reference Allele|
|ALT |	Alternate Allele|
| AF | Minimac3 output Alternate Allele Frequency|
| MAF | Minimac3 output of Minor Allele Frequency|
|SAMP_FREQ_ALT |	Alternate Allele frequency in sample being tested|
|SAMP_MAF |	Minor allele frequency in sample being tested|
| R2 | Imputation R2 score (minimac3 $R^2$)|
| ER2 | Minimac3 ouput empirical $R^2$ |
|PVALUE |	P-value of single SNP or interaction term|
|HR |	Hazard Ratio (HR)|
|HR_lowerCI |	Lower bound 95% CI of HR|
|HR_upperCI |	Upper bound 95% CI of HR|
|COEF |	Estimated coefficient of SNP|
|SE.COEF |	Standard error of coefficient estimate|
|Z |	Z-statistic|
|N |	Number of individuals in sample being tested|
|NEVENT |	Number of events that occurred in sample being tested|
