---
title: "Review and testing of gwasurvivr: an R package for genome-wide
survival analysis"

author:
- Nadia
- Ivy
- Gabriel
- Piotr

output:
  pdf_document: default
  html_document:
    df_print: paged
date: "2023-10-05"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

#for installing the packages

#install.packages("tinytex")
#tinytex::install_tinytex()
#if (!requireNamespace("BiocManager", quietly=TRUE))
#    install.packages("BiocManager")
#BiocManager::install("gwasurvivr", version = "devel")
#install.packages("devtools")
#devtools::install_github("kupietz/kableExtra")

library(gwasurvivr)
options("gwasurvivr.cores"=4) 
library(kableExtra)
```

## Introduction

Survival analysis has an important place in biomedical research, facilitating the exploration of time-to-event outcomes such as mortality or relapse. 

An essential component in exploring the genetic basis of diseases involves investigating Single Nucleotide Polymorphisms (SNPs). These occasional variations in a single letter of DNA can have a significant impact on susceptibility to certain diseases or on response to medical treatments. 

Integrating SNPs into survival analysis enables the discovery of genetic factors linked to time-to-event outcomes, shedding light on the genetic factors that influence disease progression and other critical events.

However, the major challenge is that our genome is made up of millions of SNPs, making large-scale survival analysis (GWAS) extremely complex. The existing software options for conducting such analyses are limited in several aspects (the need to interact with raw data, software not suited for survival analysis, and long execution times that hinder scalability). Consequently, researchers often face practical difficulties when conducting large-scale survival analyses.

gwasurvivr, an R/Bioconductor package was designed to surmount these challenges. This library offers a significant advancement in allowing researchers to perform survival analysis on large SNP datasets with remarkable efficiency and accuracy and with multiple file input formats such as VCF, IMPUTE2 or PLINK. 

In this paper, we thoroughly examine the functionalities of gwasurvivr and offer an extensive evaluation of its operational mechanisms and effectiveness in unraveling the genetic factors that influence disease survival.

## Methods

## Datasets

## Results

```{r benchmark, echo=F}
knitr::include_graphics("benchmark.png")
```

TO DO:
[desribe results (performance), parallelization, server implementation viability]

Michigan Imputation Server pre-phases typed genotypes using HAPI-UR, SHAPEIT, or EAGLE (default is EAGLE2), imputes using Minimac3 imputation engine and outputs Blocked GNU Zip Format VCF files (`.vcf.gz`). These `vcf.gz` files with addition of `.txt` files representing phenotype are used as an input for cox regression in `gwasurvivr` package.

The exemplary data can be extracted from the package using `system.file` function in according manner.

```{r data loading, echo=T}
vcf.file <- system.file(package="gwasurvivr",
                        "extdata", 
                        "michigan.chr14.dose.vcf.gz")
pheno.fl <- system.file(package="gwasurvivr",
                        "extdata", 
                        "simulated_pheno.txt")
```

Simulated phenotype file used as covariate file during regression can be represented in the table as shown below.

```{r data showcase, echo=F}
pheno.file <- read.table(pheno.fl,
                         sep=" ", 
                         header=TRUE,
                         stringsAsFactors = FALSE)
head(pheno.file) %>% kable(booktabs=F, linesep = "") %>% kable_styling(latex_options = c("HOLD_position"))

#decoding sex into binary format
pheno.file$SexFemale <- ifelse(pheno.file$sex=="female", 1L, 0L)
```

Now using phenotype file as covariate file and given loaded VCF file we can run `michiganCoxSurv` function, which is a wrapper for Cox regression model (before running the function, `sex` column was encoded into binary format).  

```{r cox regression, echo=T, eval=F, cache=T}
michiganCoxSurv(vcf.file=vcf.file,
                covariate.file=pheno.file,
                id.column="ID_2",
                time.to.event="time",
                event="event",
                covariates=c("age", "SexFemale", "DrugTxYes"),
                inter.term=NULL, #interaction term inclusion
                print.covs="only", #defines printing of covariates' statistics
                out.file="michigan_only",
                r2.filter=0.3, #imputation quality score filter
                maf.filter=0.005, #filter for minor allele frequency
                chunk.size=100, #number of variants to proceed per thread
                verbose=F,
                clusterObj=NULL) #for setting up cluster for computations
```

Functions saves the outputed model with the `.coxph` extension as a seperate file as well as SNPs removed (due to low variance or user defined thresholds) in the `.snps_removed` file. Accessed results of the performed regression are showcased below (`print covs = "only"` was chosen as a printing option, which omits some covariates).

```{r coxph results, echo=F}
michigan_only <- read.table("michigan_only.coxph", sep="\t", header=TRUE, stringsAsFactors = FALSE) %>% t()
kable(michigan_only, linesep = "") %>% kable_styling(latex_options = c("HOLD_position"))
```

To decipher most column names and extract knowledge from the output table in the appendix section explains the meaning of certain variables. 

Column PVALUE holds results for significance test of the SNP covariates (rest of the results is omitted by the printing option). It can be stated that on level of significance 0.05 SNP covariates (all pvalues above 0.05) can be deemed as insignificant as $H_0: coef_i = 0$. Drawn conclusion seems sensible as phenotype file was generated randomly.  

Package offers also adding SNP covariate interactions during the modelling. Setting option `inter.term` to name(s) of the variable(s), in our case `DrugTxYes`, will include these interactions during training of the regression models. Table with all (`print.covs="all"`) information about covariates is attached in the appendix. Additionally pvalues for the significance test, hazard ratios and confidence intervals for them, Z scores, coefficents and their standard errors can be obtained for all the covariates. 

Analyzing the full table of the results gives us a whole picture about performance of the covariates. Looking again at pvalues, now it's clear that only `age` can be deemed significant in this statistical inference framework.

```{r coxph interactions, echo=F, cache=T}
michiganCoxSurv(vcf.file=vcf.file,
                covariate.file=pheno.file,
                id.column="ID_2",
                time.to.event="time",
                event="event",
                covariates=c("age", "SexFemale", "DrugTxYes"),
                inter.term="DrugTxYes",
                print.covs="all",
                out.file="michigan_intx_some",
                r2.filter=0.3,
                maf.filter=0.005,
                chunk.size=100,
                verbose=F,
                clusterObj=NULL)
```

Creators of the package included three more modifications of Cox proportional hazard regression for different data types. 

`sangerCoxSurv`is the function which handles data formats obtained from Sanger Imputation Server, which pre-phases typed genotypes using either SHAPEIT or EAGLE, imputes genotypes using PBWT algorithm and outputs a .vcf.gz file for each chromosome. This cox regression wrapper allows the user to filter on info score (imputation quality metric) and minor allele frequency from the reference panel used for imputation using `RefPanelAF` as the input arguement for `maf.filter`. Users are also provided with the sample minor allele frequency in the output file. Syntax of the function and overall use cases are the same as in previously described `michiganCoxSurv`.

Another format which is handled by the package is IMPUTE2 generated output. IMPUTE2 is a genotype imputation and haplotype phasing program. It outputs 6 files for each chromosome chunk imputed, but only 2 of these files are required for analyses using `gwasurviv`. Loading and pre-processing of the genetic and phenotypic data is conducted as before. To perform survival analysis using IMPUTE2 the function arguments are very similar to `michiganCoxSurv` and `sangerCoxSurv`, however the function now takes a chromosome argument. This is needed to properly annotate the file output with the chromosome that these SNPs are in. Moreover, function `gdsCoxSurv` allows to performs survival analysis on genetic data, which was converted from IMPUTE2 format to GDS format.

Additionaly, authors are also mentioning wrapper `plinkCoxSurv`, which handles data in PLINK format (`.BED`, `.BIM` and `.FAM` files). Pre-processing, syntax and usage of the function is identical to previously desribed solutions. 

## Conclusions

`gwasurvivr`, an R package tailored for the analysis of survival outcomes in Genome-Wide Association Studies (GWAS), offers several notable advantages. It seamlessly integrates GWAS results with survival analysis, making it an accessible tool for researchers seeking to explore the genetic variants' influence on survival outcomes. 

Its seems user-friendly at first glance, as use cases are described in the vignette, but syntax and storage of the output in the external file makes it more gimmicky. Moreover, output is just a table with some metrics not a proper coxph object, which makes it harder to do inference and integrate it with different software concerning survival analysis.  

One of its main strengths is inclusion of handling diverse data format used for survival analysis inference, but then no visualization tools were added on top of cox regression wrappers.

Capacity to handle extensive datasets was the main focus of the developers. The goal was clearly achieved as the package is overall fast, has premium options for fast computing, parallelization and setting up clusters to speed up the fitting process.   

In conclusion, `gwasurvivr` is a viable options for researchers trying to incorporate genom-wide studies into survival analysis framework due to fast compuation time and fairly easy syntax. However, it has some limitations, primarily focusing on compatibility with other survival analysis framework and requiring a working knowledge of R, data preparation, and potentially resource-intensive computations. Users should also keep an eye on updates and maintenance, as with any R package.

## Appendix

|   |   |
|---|---|
| RSID  | SNP ID | 	
| TYPED | Imputation status: TRUE (SNP IS TYPED)/FALSE (SNP IS IMPUTED)|
|CHR |	Chromosome number |
|POS |	Genomic Position (BP)|
|REF |	Reference Allele|
|ALT |	Alternate Allele|
| AF | Minimac3 output Alternate Allele Frequency|
| MAF | Minimac3 output of Minor Allele Frequency|
|SAMP_FREQ_ALT |	Alternate Allele frequency in sample being tested|
|SAMP_MAF |	Minor allele frequency in sample being tested|
| R2 | Imputation R2 score (minimac3 $R^2$)|
| ER2 | Minimac3 ouput empirical $R^2$ |
|PVALUE |	P-value of single SNP or interaction term|
|HR |	Hazard Ratio (HR)|
|HR_lowerCI |	Lower bound 95% CI of HR|
|HR_upperCI |	Upper bound 95% CI of HR|
|COEF |	Estimated coefficient of SNP|
|SE.COEF |	Standard error of coefficient estimate|
|Z |	Z-statistic|
|N |	Number of individuals in sample being tested|
|NEVENT |	Number of events that occurred in sample being tested|

```{r coxph results int, echo=F}
michigan_some <- read.table("michigan_intx_some.coxph", sep="\t", header=TRUE, stringsAsFactors = FALSE) %>% t()
kable(michigan_some, linesep = "") %>% kable_styling(latex_options = c("HOLD_position"))
```