---
title: "Review and testing of gwasurvivr: an R package for genome-wide
survival analysis"

author:
- Nadia
- Ivy
- Gabriel
- Piotr

output:
  pdf_document: default
  html_document:
    df_print: paged
date: "2023-10-05"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

#for installing the packages

#install.packages("tinytex")
#tinytex::install_tinytex()
#if (!requireNamespace("BiocManager", quietly=TRUE))
#    install.packages("BiocManager")
#BiocManager::install("gwasurvivr", version = "devel")
#install.packages("devtools")
#devtools::install_github("kupietz/kableExtra")

library(gwasurvivr)
options("gwasurvivr.cores"=4) 
library(kableExtra)
```

## Introduction

Survival analysis has an important place in biomedical research, facilitating the exploration of time-to-event outcomes such as mortality or relapse. 

An essential component in exploring the genetic basis of diseases involves investigating Single Nucleotide Polymorphisms (SNPs). These occasional variations in a single letter of DNA can have a significant impact on susceptibility to certain diseases or on response to medical treatments. 

Integrating SNPs into survival analysis enables the discovery of genetic factors linked to time-to-event outcomes, shedding light on the genetic factors that influence disease progression and other critical events.

However, the major challenge is that our genome is made up of millions of SNPs, making large-scale survival analysis (GWAS) extremely complex. The existing software options for conducting such analyses are limited in several aspects (the need to interact with raw data, software not suited for survival analysis, and long execution times that hinder scalability). Consequently, researchers often face practical difficulties when conducting large-scale survival analyses.

Enter gwasurvivr, an R/Bioconductor package designed to surmount these challenges. This library offers a significant advancement in allowing researchers to perform survival analysis on large SNP datasets with remarkable efficiency and accuracy and with multiple file input formats such as VCF, IMPUTE2 or PLINK. 

In this paper, we thoroughly examine the functionalities of GWASurvivr and offer an extensive evaluation of its operational mechanisms and effectiveness in unraveling the genetic factors that influence disease survival.

## Methods

## Datasets

## Results

Michigan Imputation Server pre-phases typed genotypes using HAPI-UR, SHAPEIT, or EAGLE (default is EAGLE2), imputes using Minimac3 imputation engine and outputs Blocked GNU Zip Format VCF files (`.vcf.gz`). These `vcf.gz` files are used as input for `gwasurvivr`. Datafile retreived as a survival dataframe can be represented in `R` as below.

```{r data loading, echo=FALSE}
vcf.file <- system.file(package="gwasurvivr",
                        "extdata", 
                        "michigan.chr14.dose.vcf.gz")
pheno.fl <- system.file(package="gwasurvivr",
                        "extdata", 
                        "simulated_pheno.txt")
pheno.file <- read.table(pheno.fl,
                         sep=" ", 
                         header=TRUE,
                         stringsAsFactors = FALSE)
head(pheno.file) %>% kable(booktabs=T, linesep = "") %>% kable_styling(latex_options = c("striped", "HOLD_position"))
```

Now we can run `michiganCoxSurv` wrapper for Cox regression model.

```{r cox regression, echo=T, eval=F, cache=T}
#decoding sex into binary format
pheno.file$SexFemale <- ifelse(pheno.file$sex=="female", 1L, 0L)

michiganCoxSurv(vcf.file=vcf.file,
                covariate.file=pheno.file,
                id.column="ID_2",
                time.to.event="time",
                event="event",
                covariates=c("age", "SexFemale", "DrugTxYes"),
                inter.term=NULL,
                print.covs="only",
                out.file="michigan_only",
                r2.filter=0.3,
                maf.filter=0.005,
                chunk.size=100,
                verbose=TRUE,
                clusterObj=NULL)
```

Accessed results of the performed regression are showcased below.

```{r coxph results, echo=F}
michigan_only <- read.table("michigan_only.coxph", sep="\t", header=TRUE, stringsAsFactors = FALSE) %>% t()
kable(michigan_only, booktabs=T, linesep = "") %>% kable_styling(latex_options = c("striped", "HOLD_position"))
```

## Conclusions
